### DISCUSSION TITLE
Project Structure

### DISCUSSION THREAD

**Discussion by easrng** [ScratchAddons/ScratchAddons#99] [Discussion]
The structure of this project is pretty unclear. I think there should be a file/wiki page with information about where to add code like new addon APIs. As someone starting to try to contribute, it's pretty confusing.

**Comment by WorldLanguages** [comment]
I know, I've been adding stuff as we go and I must admit files got unorganized quickly. I have many remaining wiki pages to write... There are many decisions I took without writing them down.

**Comment by WorldLanguages** [comment]
Writing here for now, so I can use your feedback for the real structure page:
`addon-api` has the classes for the addon APIs. The ones only available in the background (persistent scripts) go on `background`, the ones only available in userscripts go on `content-script`, and common ones go into `common`.
These classes are initialized once per addon, but if an addon has multiple persistent scripts, or multiple userscripts in the same page, they share the addon object, and hence, all other objects.
Most stuff these APIs do is run simple functions, like `addon.tab.clientVersion`, but sometimes they use global state. Global state is an object that is handled by the extension's background scripts, but content scripts also have a realtime view of it. Internally, it works like this: if there's a change in global state, the background page notifies about it to the content scripts. Then, these content scripts update a special DOM element `template #scratch-addons`, specifically it's `data-global-state` attribute. Using a mutation observer, the script injected into the page (not the content script! remember extension content scripts run in a different execution context. scripts inside `content-scripts/inject` run in the same execution context as Scratch itself, and addon userscripts) identifies an update to the global state and updates `window.scratchAddons.globalState` accordingly. In conclusion, in both persistent and userscript APIs, there's a `scratchAddons.globalState` object that holds auth info, settings, etc.
There's also a similar system in place to handle events. For persistent scripts it's easy to fire an event; just go through `scratchAddons.eventTargets[eventName]` and fire an event to each. However, for userscript it's a bit harder because they're in a different execution context. For example, when settings update, it first goes to the content script, which creates a new attribute in the `template #scratch-addons` DOM element notifying the injected script of the event's details. Then, the injected script fires the events.
A similar approach is also used to make functions like `addon.account.getMsgCount()` work; when an addon requests this, the `resolve` promise function is stored inside an array inside injected script (it can't be stored anywhere else, remember, since addons run in injected script context, and you can't just move a promise resolver through execution context). Injected scripts can't communicate directly with the background page (which is in charge of requesting the message count), and injected scripts can't directly communicate with content scripts, so the DOM is used again. The injected script makes a DOM change to notify the content script about a message count request, then the content script waits for the background script to send the message count, and after it's received, the DOM is updated with the message count value. Then, the injected script goes through all messsage count pending promises and resolves them all to the value it found in the DOM.
In another matter, persistent scripts are special in a way; we want to be able to kill them. If you disable an addon with a persistent script, you don't want it listening to events or running intervals anymore. Hence, persistent scripts have special `setTimeout` and `setInterval` functions they have to use, so that the addon loader can kill the timeout if the addon is disabled. The `addon` object is also revoked through a proxy to let the browser's garbage collector remove any unused memory. For example, you can test this yourself: if you disable the Scratch Notifier addon while it's running, its arrays like `mostRecentMsgIds` are removed from memory. If the user then re enables an addon, we just run it again.
Also, I need to clarify, addons aren't executed with `eval` anymore, but with dynamic `import()`, which means addon scripts are basically JavaScript modules that export a default function, which is ran with special parameters like `addon` and `global`.
Remember I above said if an addon has multiple persistent scripts, they all share the same `addon` object? Well, they also share a `global` object. Since each addon script is a module, if you want two persistent scripts for example to share data, they could do it through the `window` object, but that would also affect other addons. Because of that, `global` exists. `global` is an empty by default object that is shared by scripts from the same addon. In persistent scripts, they all share it. In userscripts, they share it if in they're in the same page, of course.
`console` is also passed into addons so that console logs can say from what addon they came from easily.
`setTimeout`, `clearInterval`, etc. are replaced for the reasons I mentioned above.
Oof, that was a lot, and I might be missing some stuff...

**Comment by WorldLanguages** [comment]
Please provide honest feedback lol, I honestly don't know how I was able to do all of this... 

For example, so much stuff happens so that `await addon.account.getMsgCount()` resolves to the message count inside an userscript...
- Addon runs that function inside an initialized `Account` object, that was initialized by an `Addon` object. It returns the return value of `scratchAddons.methods.getMsgCount()`
- The injected extension script has that function, which returns a promise, and then adds the promise to an array. It makes a DOM attribute change so that the extension content script knows an addon needs the message count.
- The content script sees the DOM change, and sends a message to the background script that handles message counts.
- That background script calls `scratchAddons.methods.getMsgCount()` (yes, the function is named the same intentionally - that's why the `Account` class can be shared between both execution contexts), but in the extension background execution context, this function actually does the job; the `handle-msgcount.js` module checks if it was asked for the message count in the last few seconds. If not, only then it checks for the count again. After getting a result, cached or not, it checks what tab had a content script that asked for the count, and sends it a message with the new message count value.
- The content script receives that message, and updates the `data-msgcount` attribute in the DOM and sets its value to the message it received.
- The injected extension script noticed the DOM changes, goes to the "unresolved promises" array, and resolves all of them to the value it got from the DOM.
- The promise resolved, hence the addon now has the message count.

Also, now that I notice, if 2 addons ask for the message count at the same time, it might actually be requested twice, since there's no lock of any kind, in case a new request comes in while the REST API request for the previous request hasn't finished...

**Comment by WorldLanguages** [comment]
I want to write about how some stuff in the extension works, but stuff keeps changing and I keep moving the milestone :P
